src/collapse/vsprof.rs

```diff
@@ -0,0 +1,188 @@
+        // quote surrounding the function name. By splitting on this double quote, we get the
+        // function name and the remainder of the line.
+        let x: &[_] = &[',', '"'];
+        let mut split = remainder.trim_start_matches(x).splitn(2, '"');
+        match (split.next(), split.next()) {
+            (Some(function_name), Some(remainder)) => {
+                let (number_of_calls, _) = get_next_number(remainder)?;
+
+                let prev_depth = self.stack.len();
+                // There are 3 separate cases to handle regarding the depth:
+                // 1. prev_depth + 1 == depth -> a new function is called, we only need to
+                //    store the function name and the number of times it is called from the
+                //    outer function
+                // 2. prev_depth == depth -> the previous function call was a leaf node, so we
+                //    need to save the current stack and replace the top node with our node
+                //    call
+                // 3. prev_depth > depth -> the previous function call was a leaf node, so we
+                //    need to save the current stack and than we need to pop the top nodes
+                //    until the top node is our parent
+                match prev_depth.cmp(&depth) {
+                    // Case 1
+                    Ordering::Less => {
+                        self.stack
+                            .push((function_name.to_string(), number_of_calls));
+                    }
+                    // Case 2
+                    Ordering::Equal => {
+                        self.write_stack(occurences);
+                        self.stack.pop();
+                        self.stack
+                            .push((function_name.to_string(), number_of_calls));
+                    }
+                    // Case 3
+                    Ordering::Greater => {
+                        self.write_stack(occurences);
+
+                        // Pop (prev_depth - depth - 1) nodes of the stack, this leaves our
+                        // parent as the top node.
+                        for _ in 0..=prev_depth - depth {
+                            self.stack.pop();
+                        }
+
+                        self.stack
+                            .push((function_name.to_string(), number_of_calls));
+                    }
+                }
+            }
+            _ => {
+                return invalid_data_error!("Unable to parse function name from line:\n{}", line);
+            }
+        }
+
+        Ok(())
+    }
+
+    // Store the current stack in `occurences`
+    fn write_stack(&self, occurences: &mut Occurrences) {
+        let (functions, samples): (Vec<_>, Vec<_>) = self.stack.iter().cloned().unzip();
+        if let Some(number_of_calls) = samples.last() {
+            if *number_of_calls > 0 {
+                occurences.insert(functions.join(";"), *number_of_calls);
+            }
+        }
```

Comment:
```suggestion
        if let Some(nsamples) = self.stack.last().map(|(_, n)| *n) {
            if nsamples > 0 {
                let functions: Vec<_> = self.stack.iter().cloned().map(|(f, _)| f).collect();
                occurences.insert(functions.join(";"), nsamples);
            }
        }
```
This avoids doing the `unzip` work if there are no samples, and avoids allocating a `Vec` for samples, which isn't really needed.

Technically the `functions` `Vec` is wasted, since we're ultimately just collecting into a `String`, but unless you specifically want to I wouldn't worry about optimizing that :)